DOCSP-34062: tls (#309)
mongodb/docs-golang: source/fundamentals/connections/tls.txt

.. _golang-tls:

==========================
Enable TLS on a Connection
==========================

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Overview
--------

In this guide, you can learn how to connect to MongoDB instances with
the TLS security protocol.

To configure your connection to use TLS, enable

> the TLS option and provide your certificates for validation.

Author: added a TOC

   .. tab:: Connection String
      :tabid: connection string tls true

      A ``Client`` instance can connect with TLS if you set the
      ``tls`` option to ``true`` in your connection string:

      .. code-block:: go
         :emphasize-lines: 1

         uri := "mongodb://<hostname>:<port>?tls=true"
         opts := options.Client().ApplyURI(uri)
         client, _ := mongo.Connect(context.TODO(), opts)

   .. tab:: ClientOptions
      :tabid: clientoptions tls true

      A ``Client`` instance can connect with TLS if you set the
      ``TLSConfig`` field of a ``ClientOptions`` instance to an empty
      ``Config`` struct:
      
      .. code-block:: go
         :emphasize-lines: 2
         
         uri := "<connection string>"
         opts := options.Client().ApplyURI(uri).SetTLSConfig(&tls.Config{})
         client, _ := mongo.Connect(context.TODO(), opts)

.. note::
   
   If you use a DNS SRV record when connecting to MongoDB by specifying
   the ``+srv`` modification in your connection string, you enable
   TLS on your connection by default.

Within your ``Config`` instance, you can set optional
fields to configure TLS on your connection. For **testing purposes**,

> you can set the ``InsecureSkipVerify`` field to ``true``.

Author: I think this content could actually be moved to a later section.

Setting the ``InsecureSkipVerify`` field to ``true`` disables
both certificate and hostname validation.
   
.. warning::

   Specifying this option in a production environment makes
   your application insecure and potentially
   vulnerable to expired certificates and foreign processes posing
   as valid client instances.

For a full list of client options, see :ref:`golang-connection-options`.

.. _golang-configure-tls-certificates:

Configure Certificates
----------------------

> To successfully initiate a TLS request, an application must prove its

Author: THis phrasing can be changed to "complete" instead of "initiate"

identity by referencing cryptographic certificates. To connect to
MongoDB with TLS, your certificates must be stored as PEM
files.

.. important::

>    For production use, we recommend that your MongoDB deployment uses valid

Reviewer: I: ["recommend" is hypothetical -> subjunctive mood](https://www.grammarly.com/blog/english-subjunctive-what-is-it/)

```suggestion
   For production use, we recommend that your MongoDB deployment use valid
```

Author: good catch!

   certificates generated and signed by the same certificate authority.

>    For testing, it can use self-signed certificates.

Reviewer: ```suggestion
   For testing, your deployment can use self-signed certificates.
```

> The following list describes the components needed to establish

Reviewer: S: make clear where these components go

```suggestion
The following list describes the components your client must send
to the server to establish
```

> a connection with TLS:

Reviewer: ```suggestion
a TLS connection:
```

.. list-table::
   :header-rows: 1
   :widths: 30 70

   * - TLS Component
     - Description

   * - Certificate Authority (CA)
     - One or more certificate authorities to

>        trust when making a TLS connection.

Reviewer: Q: Are these the authorities the client trusts when verifying the server cert?

Author: This is the authority or authorities that have approved the certificates
The server gets a stamp of approval from the CA when evaluating the certs

Reviewer: ah, ok. i was confused because it seems suspect to let the client provide its own references, so to speak

   * - Client Certificate

>      - A digital certificate and a key that allows the server to verify the identity

Reviewer: S: Because you mention key below, and that it's often included in this file, i think you can omit it here

```suggestion
     - A digital certificate that allows the server to verify the identity
```

       of your application to establish an encrypted network connection.

   * - Certificate Key
     - The client certificate private key file. This key is often
       included within the certificate file itself.

   * - Passphrase
     - The password to decrypt the private client key if it is encrypted.

.. tip::
   
   To learn more about the PEM format, see the Wikipedia entry on

>    :wikipedia:`Privacy-Enhanced Mail <w/index.php?title=Privacy-Enhanced_Mail&oldid=1184023212>`.

Reviewer: S: move up to where you mention PEM above?

Author: didnt want to stack admonitions. I think this wiki link is marginally useful, I'll just remove.

.. _golang-client-tls-connect:

Reference Certificates in a Client
----------------------------------

You must reference your certificates in your ``ClientOptions``
object so that the server can validate them before the client connects.
We recommend that you set the ``TLSConfig`` field of your
``ClientOptions`` instance to a ``Config`` struct to configure your
TLS connection. ``Config`` structs are native to Go and allow you to keep
all your TLS options in a single reusable object.

To create a ``Config`` instance, import the ``crypto/tls`` and
``crypto/x509`` packages. Next, create a ``Config`` struct instance and
set the relevant struct fields for your configuration.

To learn more about the ``Config`` struct, see the `tls.Config API

> documentation <https://pkg.go.dev/crypto/tls#Config>`__.

Reviewer: nice 👍

.. _golang-tls-config-full-example:

Example
~~~~~~~

> This example shows how to create a ``Config`` instance and
> a ``Client`` instance that includes TLS options by performing the
> following actions:

Reviewer: S: could be cleaner to rewrite this paragraph to avoid starting every list entry with a participle

something like:

```suggestion
This example performins the
following actions to create a ``Config`` instance and
a ``Client`` instance that includes TLS options: 
```

Author: Should I shorten the code example to only include the body of main()? Or do you think showing the imports is helpful? Not usually shown in fundamentals code examples.

Reviewer: i think it's helpful. i can imagine users saying the code sample doesn't work as written.

   :wikipedia:`Transport Layer Security <w/index.php?title=Transport_Layer_Security&oldid=1184063676>`.

.. _golang-enable-tls:

Enable TLS
----------

You can enable TLS on a connection to your MongoDB instance
in one of the following ways:

- Setting the ``tls`` option to ``true`` in your connection string
- Passing an empty ``Config`` struct to the ``SetTLSConfig()``
  method when creating a ``ClientOptions`` instance

Select from the following :guilabel:`Connection String` and
:guilabel:`ClientOptions` tabs to see a corresponding code sample:

.. tabs::

   .. tab:: Connection String
      :tabid: connection string tls true

      .. code-block:: go
         :emphasize-lines: 1

         uri := "mongodb://<hostname>:<port>?tls=true"
         opts := options.Client().ApplyURI(uri)
         client, _ := mongo.Connect(context.TODO(), opts)

   .. tab:: ClientOptions
      :tabid: clientoptions tls true
      
      .. code-block:: go
         :emphasize-lines: 2
         
         uri := "<connection string>"
         opts := options.Client().ApplyURI(uri).SetTLSConfig(&tls.Config{})
         client, _ := mongo.Connect(context.TODO(), opts)

.. note::
   

>    If your connection string uses a use a DNS SRV record by including

Reviewer: I: typo here. do you think mention DNS SRV records here is useful? you could link out to the server connection docs instead.

```suggestion
   If your connection string uses a DNS SRV record by including
```

Reviewer 2: We should mention this since it's based on [user feedback](https://jira.mongodb.org/browse/DOCSP-34060). The server docs mention it, but that information is not in a format that is quickly findable.

``ClientOptions`` instance to a ``Config`` struct to configure your
TLS connection. ``Config`` structs are native to Go and allow you to keep
all your TLS options in a single reusable object.

To create a ``Config`` instance, import the ``crypto/tls`` and
``crypto/x509`` packages. Next, create a ``Config`` struct instance and
set the relevant struct fields for your configuration.

Within your ``Config`` instance, you can set optional
fields to configure TLS on your connection. For **testing purposes**,
you can set the ``InsecureSkipVerify`` field to ``true``.
   
.. warning::

   Setting the ``InsecureSkipVerify`` field to ``true`` disables
   both certificate and hostname validation.
   
   Specifying this option in a production environment makes
   your application insecure and potentially
   vulnerable to expired certificates and foreign processes posing
   as valid client instances.

To learn more about the ``Config`` struct, see the `tls.Config API
documentation <https://pkg.go.dev/crypto/tls#Config>`__.

.. _golang-tls-config-full-example:

Example
~~~~~~~

This example performs the following actions to create a ``Config``

> instance and a ``Client`` instance that includes the TLS options:

Reviewer: s: "the TLS options" stands out to me. could it be changed to "instance enabled for TLS" or "the previous TLS options" (if that's true)?
